src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
-					LOG_WARNING("Batch memory read encountered DMI error %d. "
-							"Falling back on slower reads.", status);
-					riscv_batch_free(batch);
-					result = ERROR_FAIL;
-					goto error;
-				}
-				value <<= 32;
-				value |= riscv_batch_get_dmi_read_data(batch, read_count);
-				read_count++;
-			}
-			riscv_addr_t offset = j * size;
-			buf_set_u64(buffer + offset, 0, 8 * size, value);
-			log_memory_access64(address + j * increment, value, size, true);
-		}
+	const uint32_t elements_to_read = read_memory_progbuf_inner_fill_batch(batch,
+			loop_count - index, access.element_size);
 
-		index = next_index;
+	int result = read_memory_progbuf_inner_run_and_process_batch(target, batch,
+			access, index, elements_to_read, elements_read);
+	riscv_batch_free(batch);
+	return result;
+}
 
-		riscv_batch_free(batch);
+/**
+ * read_memory_progbuf_inner_startup() must be called before calling this function
+ * with the address argument equal to curr_target_address.
+ */
+static int read_memory_progbuf_inner_ensure_forward_progress(struct target *target,
+		struct memory_access_info access, uint32_t start_index)
+{
+	LOG_TARGET_DEBUG(target,
+			"Executing one loop iteration to ensure forward progress (index=%"
+			PRIu32 ")", start_index);
+	const target_addr_t curr_target_address = access.target_address +
+		start_index * access.increment;
+	uint8_t * const curr_buffer_address = access.buffer_address +
+		start_index * access.element_size;
+	const struct memory_access_info curr_access = {
+		.buffer_address = curr_buffer_address,
+		.target_address = curr_target_address,
+		.element_size = access.element_size,
+		.increment = access.increment,
+	};
+	uint32_t elements_read;
+	if (read_memory_progbuf_inner_try_to_read(target, curr_access, &elements_read,
+			/*index*/ 0, /*loop_count*/ 1) != ERROR_OK)
+		return ERROR_FAIL;
+
+	if (elements_read != 1) {
+		assert(elements_read == 0);
+		LOG_TARGET_DEBUG(target, "Can not ensure forward progress");
+		/* FIXME: Here it would be better to retry the read and fail only if the
+		 * delay is greater then some threshold.
+		 */
+		return ERROR_FAIL;
 	}
+	return ERROR_OK;
+}
 
-	dmi_write(target, DM_ABSTRACTAUTO, 0);
+static void set_buffer_and_log_read(struct memory_access_info access,
+		uint32_t index, uint64_t value)
+{
+	uint8_t * const buffer = access.buffer_address;
+	const uint32_t size = access.element_size;
+	const uint32_t increment = access.increment;
+	const target_addr_t address = access.target_address;
+
+	assert(size <= 8);
+	buf_set_u64(buffer + index * size, 0, 8 * size, value);
+	log_memory_access64(address + index * increment, value, size,
+			/*is_read*/ true);
+}
 
-	if (count > 1) {
-		/* Read the penultimate word. */
-		uint32_t dmi_data0, dmi_data1 = 0;
-		if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK)
-			return ERROR_FAIL;
-		if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK)
+static int read_word_from_dmi(struct target *target,
```

Comment:
I'd recommend to call this "read_word_from_data_regs()", as opposed to "..from_dmi()".

```suggestion
static int read_word_from_data_regs(struct target *target,
```

Or, even better, I think you can use the existing `read_abstract_arg()` function.

Reply:
Thanks for the great suggestion! Addressed.